src/Microsoft.ML.AutoML.Interactive/NotebookMonitor.cs

```diff
@@ -0,0 +1,94 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using Microsoft.DotNet.Interactive;
+using System.Collections.Generic;
+using Microsoft.Data.Analysis;
+using System;
+using System.Threading.Tasks;
+using System.Text.Json;
+using System.Threading;
+
+namespace Microsoft.ML.AutoML
+{
+    public class NotebookMonitor : IMonitor
+    {
+        private DisplayedValue _valueToUpdate;
+        private DateTime _lastUpdate = DateTime.MinValue;
+
+        public TrialResult BestTrial { get; set; }
+        public TrialResult MostRecentTrial { get; set; }
+        public TrialSettings ActiveTrial { get; set; }
+        public List<TrialResult> CompletedTrials { get; set; }
+        public DataFrame TrialData { get; set; }
+
+        public NotebookMonitor()
+        {
+            CompletedTrials = new List<TrialResult>();
+            TrialData = new DataFrame(new PrimitiveDataFrameColumn<int>("Trial"), new PrimitiveDataFrameColumn<float>("Metric"), new StringDataFrameColumn("Trainer"), new StringDataFrameColumn("Parameters"));
+        }
+
+        public void ReportBestTrial(TrialResult result)
+        {
+            BestTrial = result;
+            Update();
+        }
+
+        public void ReportCompletedTrial(TrialResult result)
+        {
+            MostRecentTrial = result;
+            CompletedTrials.Add(result);
+
+            var activeRunParam = JsonSerializer.Serialize(result.TrialSettings.Parameter, new JsonSerializerOptions() { WriteIndented = false, });
+
+            TrialData.Append(new List<KeyValuePair<string, object>>()
+            {
+                new KeyValuePair<string, object>("Trial",result.TrialSettings.TrialId),
+                new KeyValuePair<string, object>("Metric", result.Metric),
+                new KeyValuePair<string, object>("Trainer",result.TrialSettings.Pipeline.ToString().Replace("Unknown=>","")),
+                new KeyValuePair<string, object>("Parameters",activeRunParam),
+            }, true);
+            Update();
+        }
+
+        public void ReportFailTrial(TrialResult result)
+        {
+            // TODO figure out what to do with failed trials.
+            Update();
+        }
+
+        public void ReportRunningTrial(TrialSettings setting)
+        {
+            ActiveTrial = setting;
+            Update();
+        }
+
+        private int _updatePending = 0;
+        public void Update()
+        {
+            Task.Run(async () =>
+            {
+                if (Interlocked.CompareExchange(ref _updatePending, 1, 0) == 0) // _updatePending is int initialized with 0
+                {
+                    DateTime n = DateTime.UtcNow;
+                    if (n - _lastUpdate < TimeSpan.FromSeconds(5))
+                    {
+                        await Task.Delay(n - _lastUpdate);
```

Comment:
No worries - I should have caught it :). That's what I get for coding too late.

Reply:
Moved to:
https://github.com/dotnet/machinelearning/pull/6243/files#diff-38e7b5c79365be27915c2761266d21f78921c344992732ed2f0b2a4360f4ddc2R34

Reply:
I had a pending refactor ... so I included the refactor with this change.

Reply:
Original code had another bug ... it would do an double update because _lastUpdate was being set to `n` instead of when the actual update was `UtcNow`.

Reply:
Would that make a difference? does the update take long time to finish which make a significant difference between n and `UtcNow`? Your update is good anyway. I am just wondering.

Reply:
When the minDelay is small it matters more. The update can take 300-1200ms.

Reply:
Also `n` was before delay ... `n` could have nearly been `minDelay` behind `UtcNow`